include/micm/util/cublas_handle_singleton.hpp

```diff
@@ -0,0 +1,72 @@
+#include <memory>
+#include <mutex>
+#include <map>
+#include <cuda_runtime.h>
+#include <cublas_v2.h>
+#include <micm/util/cuda_util.cuh>
+
+namespace micm
+{
+  class CublasHandleSingleton {
+    public:
+      // Delete copy constructor and assignment operator to prevent copies
+      CublasHandleSingleton(const CublasHandleSingleton&) = delete;
+      CublasHandleSingleton& operator=(const CublasHandleSingleton&) = delete;
+
+      // Get the static instance of CublasHandleSingleton class
+      static CublasHandleSingleton& GetInstance(int device_id)
+      {
+        if (auto search = cublas_handle_map_.find(device_id); search == cublas_handle_map_.end())
+        {
+          std::lock_guard<std::mutex> lock(GetMutex()); // no cublas handle if found; lock the mutex and generate a new cublas handle below
+        }
```

Comment:
Placing the lock inside of the if will immediately grab the lock but then let it go once the if block ends because the lock will go out of scope. So we should just grab the lock every time 

```suggestion  
        std::lock_guard<std::mutex> lock(GetMutex()); // no cublas handle if found; lock the mutex and generate a new cublas handle below
```

Reply:
Thanks for catching it. Fixed!